### PR TITLE
[DEE-457] "PropertyDrift" and "TrainTestPerformance" checks

### PR DESCRIPTION
**Pull request by yromanyshyn**

### CODE REVIEW

File: deepchecks/utils/abstracts/train_test_performace.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+__all__ = ['TrainTestPerformanceAbstract']
+
+
+class TrainTestPerformanceAbstract:
+    """Base functionality for some train-test performance checks."""
+
+    @classmethod
+    @abc.abstractmethod
+    def _default_per_class_scorers(cls) -> t.Mapping[str, str]:
+        raise NotImplementedError()
```
Comment (Nadav-Barak) [inline deepchecks/utils/abstracts/train_test_performace.py:31]: Why is this needed? 
In general the more prefered flow is to get the scorers list from context.get_scorers(self.scorers, use_avg_defaults=False) - as it was in the tabular sub package

File: deepchecks/nlp/checks/model_evaluation/train_test_performance.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        self.n_samples = n_samples
+        self.random_state = random_state
+
+    @classmethod
+    def _default_per_class_scorers(cls) -> t.Dict[str, str]:
+        # TODO:
+        return {}
+
+    def run_logic(self, context: Context) -> CheckResult:
+        """Run check."""
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/model_evaluation/train_test_performance.py:100]: Try to move more of this section into the Abstract - similar to what happening in PredictionDriftAbstract - a function the return CheckResult
Optimal - send scorers and the dataset and do everything there
Minimal - create the display in the abstract (send the scorers results)

File: deepchecks/nlp/checks/train_test_validation/property_drift.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        n_top_properties: int = 5,
+        margin_quantile_filter: float = 0.025,
+        max_num_categories_for_drift: t.Optional[int] = None,
+        min_category_size_ratio: float = 0.01,
+        max_num_categories_for_display: int = 10,
+        show_categories_by: str = 'largest_difference',
+        numerical_drift_method: str = 'KS',
+        categorical_drift_method: str = 'cramers_v',
+        ignore_na: bool = True,
+        aggregation_method: t.Optional[str] = 'l3_weighted',
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/property_drift.py:114]: remove the aggregation_method argument

File: deepchecks/nlp/checks/train_test_validation/property_drift.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        min_samples: int = 10,
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/property_drift.py:115]: change min samples to 100

File: deepchecks/nlp/checks/train_test_validation/property_drift.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        self.categorical_drift_method = categorical_drift_method
+        self.ignore_na = ignore_na
+        self.aggregation_method = aggregation_method
+        self.min_samples = min_samples
+        self.n_samples = n_samples
+        self.random_state = random_state
+
+    def run_logic(self, context: Context) -> CheckResult:
+        """Run check."""
+        train = t.cast(TextData, context.train)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/property_drift.py:139]: why is the cast needed? can you do in the context property call?

File: deepchecks/nlp/checks/train_test_validation/property_drift.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            columns=self.properties,
+            ignore_columns=self.ignore_properties
+        )
+        test_properties = select_from_dataframe(
+            test.properties,
+            columns=self.properties,
+            ignore_columns=self.ignore_properties
+        )
+
+        columns = set(train_properties.columns).intersection(set(test_properties.columns))
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/property_drift.py:153]: add a function called in creation of context that validate if the train & test have the same properties + properties type and if not display a warning stating only common properties will be used in train-test checks.

Comment (Nadav-Barak) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py:153]: best if can be a generic structure like in validate_length_and_calculate_column_types so in the future we will also validate metadata and embeddings in the same manner

File: deepchecks/nlp/checks/train_test_validation/property_drift.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        cat_columns = set([*train.categorical_properties, *test.categorical_properties])
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/property_drift.py:154]: only use columns that are common to both train and test properties

File: deepchecks/nlp/checks/train_test_validation/property_drift.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+        results = {}
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/train_test_validation/property_drift.py:155]: All code below this, can they be put under a abstract class? Similar to what you did in the TrainTestPerformance

I mean here a common abstract class that will be the parent of this class and of FeatureDrift tabular check

File: tests/nlp/checks/train_test_validation/property_drift_test.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            'Max Word Length': has_entries({'Drift score': 0.18000000000000005, 'Method': 'Kolmogorov-Smirnov'}),
+            'Average Word Length': has_entries({'Drift score': 0.1, 'Method': 'Kolmogorov-Smirnov'}),
+            '% Special Characters': has_entries({'Drift score': 0.16000000000000003, 'Method': 'Kolmogorov-Smirnov'}),
+            'Text Length': has_entries({'Drift score': 0.30000000000000004, 'Method': 'Kolmogorov-Smirnov'}),
+            'Subjectivity': has_entries({'Drift score': 0.14, 'Method': 'Kolmogorov-Smirnov'}),
+            'Sentiment': has_entries({'Drift score': 0.08000000000000007, 'Method': 'Kolmogorov-Smirnov'})
+        }))  # type: ignore
+
+
+class TestLabelClassification:
```
Comment (Nadav-Barak) [inline tests/nlp/checks/train_test_validation/property_drift_test.py:132]: multilabel

File: tests/nlp/checks/train_test_validation/property_drift_test.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+            raw_text=[
+                random.choice(['I think therefore I am', 'I am therefore I think', 'I am'])
+                for _ in range(20)
+            ],
+            label=[
+                random.choice([[0, 0, 1], [1, 1, 0], [0, 1, 0]])
+                for _ in range(20)
+            ],
+            task_type='text_classification'
+        )
```
Comment (Nadav-Barak) [inline tests/nlp/checks/train_test_validation/property_drift_test.py:145]: Put this one at context as 'dummy multilabel dataset'

File: deepchecks/nlp/checks/train_test_validation/property_drift.py (hunk 3 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py]: fixed

File: deepchecks/utils/abstracts/train_test_performace.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/utils/abstracts/train_test_performace.py:31]: because `add_condition_class_performance_imbalance_ratio_less_than` needs access to the default scorers

```
def add_condition_class_performance_imbalance_ratio_less_than(
    self: Self,
    threshold: float = 0.3,
    score: t.Optional[str] = None
) -> Self:
    if score is None:
        score = next(iter(self._default_per_class_scorers()))

```

File: deepchecks/nlp/checks/train_test_validation/property_drift.py (hunk 5 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py:133]: to say to the static type analyzer that the `context.train` is of type `TextData`

currently, the static type analyzer infers those properties to be of type `Never`, which means that code cannot and will not be reached, it is because `nlp.context.Context` type inherences those properties from `BaseContext` in which they are defined as `None`

```
class BaseContext(ABC):
    """Base class for contexts."""

    _train = None
    _test = None
    
    ...

    @property
    def train(self):
        """Return train if exists, otherwise raise error."""
        if self._train is None:
            raise DeepchecksNotSupportedError('Check is irrelevant for Datasets without train dataset')
        return self._train

```

Never


> can you do in the context property call?

do you suggest overriding `train` and `test` properties in the nlp `Context` type?

Comment (yromanyshyn) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py:133]: >can you do in the context property call?

or we can make `BaseContext` generic?

File: tests/nlp/checks/train_test_validation/property_drift_test.py (hunk 9 above)
Comment (yromanyshyn) [reply tests/nlp/checks/train_test_validation/property_drift_test.py]: fixed

File: deepchecks/nlp/checks/train_test_validation/property_drift.py (hunk 7 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py]: fixed, now use only common columns that have the same type, other columns are igored

File: deepchecks/nlp/checks/train_test_validation/property_drift.py (hunk 6 above)
Comment (yromanyshyn) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py]: added

File: tests/nlp/checks/train_test_validation/property_drift_test.py (hunk 10 above)
Comment (yromanyshyn) [reply tests/nlp/checks/train_test_validation/property_drift_test.py:145]: fixed

File: deepchecks/nlp/checks/train_test_validation/property_drift.py (hunk 5 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/train_test_validation/property_drift.py:134]: I think the solution is too complicated so we can just leave it as is.
I thought we can do something simple like adding return type declaration to the property

File: deepchecks/nlp/checks/model_evaluation/train_test_performance.py (hunk 2 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/model_evaluation/train_test_performance.py:98]: Yea ok

File: deepchecks/utils/abstracts/train_test_performace.py (hunk 1 above)
Comment (Nadav-Barak) [reply deepchecks/utils/abstracts/train_test_performace.py:34]: wow that is some bad implementation.
Change the condition so it will be as in deepchecks/vision/checks/model_evaluation/class_performance.py
i.e the scorer name is received as input from the user

Comment (yromanyshyn) [reply deepchecks/utils/abstracts/train_test_performace.py]: oh, I see why it was done so, to not change the function signature, to not break backward compatibility

anyway, I do not see this as a problematic change/

File: deepchecks/nlp/context.py [hunk 11]
```diff
@@ -326,14 +329,45 @@ def __init__(
...
             if test_dataset.name is None:
                 test_dataset.name = 'Test'
+
         # If both dataset, validate they fit each other
         if train_dataset and test_dataset:
-            if test_dataset.has_label() and train_dataset.has_label() and not \
-                    train_dataset.validate_textdata_compatibility(test_dataset):
+            if (
+                test_dataset.has_label()
+                and train_dataset.has_label()
```
Comment (Nadav-Barak) [inline deepchecks/nlp/context.py:337]: why is this needed? validate_textdata_compatibility should be able to work regardless is labels are present or not

File: deepchecks/nlp/context.py [hunk 12]
```diff
@@ -326,14 +329,45 @@ def __init__(
...
+                and not train_dataset.validate_textdata_compatibility(test_dataset)
+            ):
                 raise DatasetValidationError('train_dataset and test_dataset must share the same label and task type')
+
+            if (
+                train_dataset._properties is not None
+                and train_dataset._cat_properties is not None
+                and test_dataset._properties is not None
+                and test_dataset._cat_properties is not None
+            ):
```
Comment (Nadav-Barak) [inline deepchecks/nlp/context.py:347]: The if is redundant, the compare method should be able to menage situations where some of those parameters are none and the rest are not
e.x only test_dataset._cat_properties is none

File: deepchecks/utils/abstracts/feature_drift.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            sorted_by = 'drift score'
+            key = lambda col: results[col]['Drift score'] or 0
+            columns_order = sorted(results.keys(), key=key, reverse=True)[:self.n_top_columns]
+
+        headnote = [
+            textwrap.dedent(
+                f"""
+                <span>
+                The Drift score is a measure for the difference between two distributions, in this check - the test
+                and train distributions.<br> The check shows the drift score and distributions for the features, sorted
```
Comment (Nadav-Barak) [inline deepchecks/utils/abstracts/feature_drift.py:128]: replace features with a argument received externally stating wether it is feature or property - 
same as data_table_name argument in validate_length_and_calculate_column_types

**Review by Nadav-Barak — APPROVED** [review]

File: deepchecks/utils/abstracts/feature_drift.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+                by {sorted_by} and showing only the top {self.n_top_columns} features, according to {sorted_by}.
+                </span>
+                """
+            ),
+            get_drift_plot_sidenote(
+                self.max_num_categories_for_display,
+                self.show_categories_by
+            ),
+            'If available, the plot titles also show the feature importance (FI) rank'
```
Comment (Nadav-Barak) [inline deepchecks/utils/abstracts/feature_drift.py:137]: remove this text for properties

Comment (yromanyshyn) [reply deepchecks/utils/abstracts/feature_drift.py]: fixed

File: deepchecks/utils/abstracts/feature_drift.py (hunk 13 above)
Comment (yromanyshyn) [reply deepchecks/utils/abstracts/feature_drift.py]: fixed

File: deepchecks/nlp/context.py (hunk 12 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py:347]: > the compare method should be able to menage situations where some of those parameters are none and the rest are not

hm/ I disagree, IMO, you are suggesting leaking `Context` specific logic into the `compare_dataframes` function, and that is not good

a task of the  `compare_dataframes` function is two compare two data frames, it should not care about anything more, making some of its parameters optional just will overcomplicate this function and makes it more confusing to understand

> e.x only test_dataset._cat_properties is none

sorry, do not understand/
`_properties` also can be none

its definition
```
_properties: t.Optional[pd.DataFrame] = None
```

File: deepchecks/nlp/context.py (hunk 11 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py:337]: hm, I do not know/ I did not add this, I just moved each part of the condition exp on the new line

File: deepchecks/nlp/context.py (hunk 12 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/context.py:347]: Main requirement - 
if one text data has 5 properties  with 1 of them being categorical and the other has 4 properties with 0 categorical (meaning _cat_properties is None) we still want the functionalities to work

Comment (Nadav-Barak) [reply deepchecks/nlp/context.py:347]: Specifically IMO the logic to handle this case should be inside the compare_dataframes function

File: deepchecks/nlp/context.py (hunk 11 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/context.py:337]: ok yea i see

**Review by Nadav-Barak — APPROVED** [review]

File: deepchecks/nlp/context.py (hunk 12 above)
Comment (yromanyshyn) [reply deepchecks/nlp/context.py]: >if one text data has 5 properties with 1 of them being categorical and the other has 4 properties with 0 categorical (meaning _cat_properties is None) we still want the functionalities to work

oh, yeap, did not think about that, thanks!

>Specifically IMO the logic to handle this case should be inside the compare_dataframes function

moving that `if` statement into the `compare_dataframe` function just means that it will appear just a little bit later in the code

```
# NOW >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

if (
    train_dataset._properties is not None
    and test_dataset._properties is not None
):
    result = compare_dataframes(
        train=train_dataset._properties,
        test=test_dataset._properties,
        train_categorical_columns=train_dataset._cat_properties,
        test_categorical_columns=test_dataset._cat_properties
    )
    self._common_datasets_properties = result.common
    if result.difference is not None:
        get_logger().warning(
            'Train and Test datasets have different properties, '

# SUGGESTION >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

# returns none if one of the dataframes is none
result = compare_dataframes(
    train=train_dataset._properties,
    test=test_dataset._properties,
    train_categorical_columns=train_dataset._cat_properties,
    test_categorical_columns=test_dataset._cat_properties
)
if result is not None:
    self._common_datasets_properties = result.common
    if result.difference is not None:
        get_logger().warning(
            'Train and Test datasets have different properties, '
```
